### ISSUE TITLE
eBPF event loop blockage finder

### ISSUE THREAD

**Issue by kvakil** [nodejs/diagnostics#569]
Hi -- I'd like to share an eBPF use-case which I found quite useful in
my day job. (Please let me know if there is a better forum to do this.)

We were experiencing long (10s+) event loop blockages which was
affecting our performance. We were alerted to this issue by
[node-blocked][nb].

[nb]: https://github.com/tj/node-blocked

We had two initial ideas:

1. CPU profiles: but the overhead felt too high especially since the
   blockages only happened rarely.
1. `async_hooks` (specifically [blocked-at][ba]): but the overhead was
   unacceptable.

[ba]: https://github.com/naugtur/blocked-at

The solution we landed on used eBPF, particularly bpftrace:

```bpftrace
/* Whenever any thread enters uv__run_timers, record the current time
   in nanoseconds in a map. */
u:NODE_PATH:uv__run_timers { @[tid] = nsecs; }

/* Whenever any thread returns from uv__run_check, clear its time from
   the map. */
ur:NODE_PATH:uv__run_check /@[tid]/ { delete(@[tid]); }

/* 99 times a second, check if any running thread has been blocked
   for longer than 10 seconds. If so, take a core dump and stop
   this script. */
p:hz:99 /@[tid]/ {
    if (nsecs - @[tid] > 10000000000) {
        system("gcore %d", pid);
        exit();
    }
}
```

We ran this script on a bunch of machines, and eventually it spit out a
coredump. We opened the coredump with [llnode][lln] and found the cause
via `v8 backtrace`.

[lln]: https://github.com/nodejs/llnode

## Questions for this group

* I will also create a separate issue about how llnode is no longer
  supported, but I think this is still useful functionality. For
  example, you can use it to get histograms of event loop blockages
  which is independently useful for workload characterization.

* On Node's side, it would be nicer if event loop stages were exposed as
  stable tracepoints instead of uprobes. This would make it easier for
  people to package similar tools.

* One could also imagine a weaker version of this functionality being
  built-in to NodeJS: collecting a Javascript backtrace when the event
  loop has currently been blocked for too long. From talking with other
  engineers, I've heard that attributing event loop blockages is a
  common problem when running NodeJS at scale. Is there interest in
  having this in NodeJS core?

**Comment by kvakil** [comment]
> * One could also imagine a weaker version of this functionality being
>       built-in to NodeJS: collecting a Javascript backtrace when the event
>       loop has currently been blocked for too long. From talking with other
>       engineers, I've heard that attributing event loop blockages is a
>       common problem when running NodeJS at scale. Is there interest in
>       having this in NodeJS core?

Here is an PoC to show what I mean: https://git.sr.ht/~kvakil/fast-blocked-at/tree
See `native.cc` for the native code and `test.js` for the example usage.

**Comment by RafaelGSS** [comment]
> I will also create a separate issue about how `llnode` is no longer
supported, but I think this is still useful functionality. For
example, you can use it to get histograms of event loop blockages
which is independently useful for workload characterization.

For `llnode`, see: https://github.com/nodejs/node/pull/43289

> On Node's side, it would be nicer if event loop stages were exposed as
stable tracepoints instead of uprobes. This would make it easier for
people to package similar tools.

Yes, that's a good improvement indeed.

---

Basically, in our last WG Meeting, I said to wait a while regarding the eBPF support, because I'm definitely interested in helping on that. Unfortunately, I had no time recently to investigate. But, soon as possible I'll try to collect what Node.js would need to support it.

Feel free to investigate too, any help is welcome! I'd say, once it doesn't hurt the Node.js Performance, it will be definitely accepted by the team.

**Comment by theanarkh** [comment]
I have written a [demo](https://github.com/theanarkh/libbpf-code/blob/main/src/uv_uprobe.bpf.c) by ebpf before to count the cost of every phase of libuv. I think ebpf is very useful for solving problem.

**Comment by RafaelGSS** [comment]
> I have written a [demo](https://github.com/theanarkh/libbpf-code/blob/main/src/uv_uprobe.bpf.c) by ebpf before to count the cost of every phase of libuv. I think ebpf is very useful for solving problem.

It seems private. 404.

**Comment by theanarkh** [comment]
> > I have written a [demo](https://github.com/theanarkh/libbpf-code/blob/main/src/uv_uprobe.bpf.c) by ebpf before to count the cost of every phase of libuv. I think ebpf is very useful for solving problem.
> 
> It seems private. 404.

Oh ! sorry, fixed.

**Comment by kvakil** [comment]
> I have written a [demo](https://github.com/theanarkh/libbpf-code/blob/main/src/uv_uprobe.c) by ebpf before to count the cost of every phase of libuv. I think ebpf is very useful for solving problem.

Thanks for the confirmation that this approach has been useful to
others. [The accompanying blog post][ap] shows one of the big problems
with uprobes: there are cases (usually inlining) where attaching the
probe fails. A stable tracepoint interface would be better from that
perspective.

[ap]: https://blog.csdn.net/THEANARKH/article/details/122006904 

> > On Node's side, it would be nicer if event loop stages were exposed as
> > stable tracepoints instead of uprobes. This would make it easier for
> > people to package similar tools.
> 
> Yes, that's a good improvement indeed.
> 
> Basically, in our last WG Meeting, I said to wait a while regarding the eBPF support, because I'm definitely interested in helping on that. Unfortunately, I had no time recently to investigate. But, soon as possible I'll try to collect what Node.js would need to support it.

I appreciate the insight. I think that the next step here is to
re-implement ebPF probes in libuv, e.g. through libuv/libuv#2209. There
is also some previous discussion on whether native tracepoints are
useful in this WG (#10, #163).

**Comment by theanarkh** [comment]
Yes, i think this is the problem of uprobe/kprobe that the name of function may change. I'm looking forward to seeing Libuv support these capabilities(trace the event loop).

**Comment by kvakil** [comment]
It turns out Node.js recently removed the dtrace probes:
nodejs/node#43652. It seems sort of like native tracing capabilities
are not considered "first-class" in the node ecosystem, so pursuing this
further seems a little futile. :\

**Comment by JiaHuann** [comment]
> Yes, i think this is the problem of uprobe/kprobe that the name of function may change. I'm looking forward to seeing Libuv support these capabilities(trace the event loop).

Actually I want to do this job as well,the  ["metrics" API](https://github.com/libuv/libuv/blob/v1.x/docs/src/metrics.rst) to have some overhead on additional "epoll_wait", [libuv-issues-3937](https://github.com/libuv/libuv/issues/3937). It is a great idea to use ebpf to instead it.But i am not sure how much work we need to do.
